cl/beacon/handler/block_production.go

```diff
@@ -487,3 +717,124 @@ func (a *ApiHandler) storeBlockAndBlobs(ctx context.Context, block *cltypes.Sign
 
 	return a.forkchoiceStore.OnBlock(ctx, block, true, false, false)
 }
+
+type attestationCandidate struct {
+	attestation *solid.Attestation
+	reward      uint64
+}
+
+func (a *ApiHandler) findBestAttestationsForBlockProduction(
+	s abstract.BeaconState,
+) *solid.ListSSZ[*solid.Attestation] {
+
+	ret := solid.NewDynamicListSSZ[*solid.Attestation](int(a.beaconChainCfg.MaxAttestations))
+	attestationCandidates := []attestationCandidate{}
+
+	for _, attestation := range a.operationsPool.AttestationsPool.Raw() {
+		if err := eth2.IsAttestationApplicable(s, attestation); err != nil {
+			continue // attestation not applicable skip
+		}
+		expectedReward, err := computeAttestationReward(s, attestation)
+		if err != nil {
+			log.Warn(
+				"[Block Production] Could not compute expected attestation reward",
+				"reason",
+				err,
+			)
+			continue
+		}
+		if expectedReward == 0 {
+			continue
+		}
+		attestationCandidates = append(attestationCandidates, attestationCandidate{
+			attestation: attestation,
+			reward:      expectedReward,
+		})
+	}
+	// Rank by reward in descending order.
+	sort.Slice(attestationCandidates, func(i, j int) bool {
+		return attestationCandidates[i].reward > attestationCandidates[j].reward
+	})
+	// Some aggregates can be supersets of existing ones so let's filter out the supersets
+	// this MAP is HashTreeRoot(AttestationData) => AggregationBits
+	aggregationBitsByAttestationData := make(map[libcommon.Hash][]byte)
+	for _, candidate := range attestationCandidates {
+		// Check if it is a superset of a pre-included attestation with higher reward
+		attestationDataRoot, err := candidate.attestation.AttestantionData().HashSSZ()
+		if err != nil {
+			log.Warn("[Block Production] Cannot compute attestation data root", "err", err)
+			continue
+		}
+		currAggregationBits, exists := aggregationBitsByAttestationData[attestationDataRoot]
+		if exists {
+			if utils.IsNonStrictSupersetBitlist(
+				currAggregationBits,
+				candidate.attestation.AggregationBits(),
+			) {
+				continue
+			}
+			utils.MergeBitlists(currAggregationBits, candidate.attestation.AggregationBits())
+		} else {
+			currAggregationBits = candidate.attestation.AggregationBits()
+		}
+		// Update the currently built superset
+		aggregationBitsByAttestationData[attestationDataRoot] = currAggregationBits
+
+		ret.Append(candidate.attestation)
+		if ret.Len() >= int(a.beaconChainCfg.MaxAttestations) {
+			break
+		}
+	}
+	return ret
+}
+
+// computeAttestationReward computes the reward for a specific attestation.
+func computeAttestationReward(
```

Comment:
is there any reference to formula of this reward computation

Reply:
answered in discord